### PR TITLE
pallet macro: always generate storage info on pallet struct

### PR DESCRIPTION
**Pull request by thiolliere**
Currently the trait `StorageInfoTrait` is implemented on the struct `Pallet` only when the attribute `generate_storage_info` is set.
But it can be handy to always generate the storage info even with partial information.

This PR makes storage implement a new trait PartialStorageInfoTrait which allows to return some partial information.
Then depending on whereas the attribute `generate_storage_info` is set, the Pallet will implement the StorageInfo using PartialStorageInfo or using StorageInfo

as a side note, the span should be improved in a follow up by https://github.com/paritytech/substrate/pull/8850
the implementation should also be improved currently we allocate some small vec way too much. But we can do in a follow-up.

cc @shawntabrizi

### CODE REVIEW

**Review by shawntabrizi — APPROVED** [review]
lgtm

File: frame/support/procedural/src/pallet/expand/pallet_struct.rs [hunk 1]
```diff
@@ -102,41 +102,58 @@ pub fn expand_pallet_struct(def: &mut Def) -> proc_macro2::TokenStream {
...
-			}
+	let storage_info_span = def.pallet_struct.generate_storage_info
+		.unwrap_or(def.pallet_struct.attr_span);
+
+	// Depending on the flag `generate_storage_info` we use partial or full storage info from
+	// storage.
+	let (
+		storage_info_trait,
+		storage_info_method,
+	) = if def.pallet_struct.generate_storage_info.is_some() {
```
Comment (KiChjang) [inline frame/support/procedural/src/pallet/expand/pallet_struct.rs:113]: You might as well compute and return the span here, no reason to separate them both into different let bindings.

**Review by KiChjang — APPROVED** [review]
Looks good overall. I think one question that I would have now is the discoverability of storage items that were implemented with `PartialStorageInfo`. This is useful when I may later on want to visit those items and implement `StorageInfo` directly, but I guess this is more of a UI issue than anything else.

File: frame/support/procedural/src/pallet/expand/pallet_struct.rs (hunk 1 above)
Comment (thiolliere) [reply frame/support/procedural/src/pallet/expand/pallet_struct.rs]: yes indeed

**Comment by thiolliere** [comment]
> Looks good overall. I think one question that I would have now is the discoverability of storage items that were implemented with `PartialStorageInfo`. This is useful when I may later on want to visit those items and implement `StorageInfo` directly, but I guess this is more of a UI issue than anything else.

The partial storage info doesn't have the max_size bound. I think when the benchmarking process fail to get proper bound for max_size it will return an error or a warning saying that the storage doesn't have valid bound for PoV size computation.

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**